File: io-hotmoka-network/src/main/java/io/hotmoka/network/internal/models/requests/NonInitialTransactionRequestModel.java
package io.hotmoka.network.internal.models.requests;

import java.math.BigInteger;

import io.hotmoka.network.internal.models.storage.StorageReferenceModel;
import io.hotmoka.network.internal.models.storage.TransactionReferenceModel;

public abstract class NonInitialTransactionRequestModel extends TransactionRequestModel {
    private String signature;
    private StorageReferenceModel caller;
    private BigInteger nonce;
    private TransactionReferenceModel classpath;
    private String chainId;
    private BigInteger gasLimit;
    private BigInteger gasPrice;

    protected final TransactionReferenceModel getClasspath() {
        return classpath;
    }

    public void setClasspath(TransactionReferenceModel classpath) {
    	System.out.println("setting classpath to " + classpath);
        this.classpath = classpath;
    }

    protected final BigInteger getNonce() {
        return nonce;
    }

    public void setNonce(BigInteger nonce) {
    	System.out.println("setting nonce to " + nonce);
        this.nonce = nonce;
    }

    protected final StorageReferenceModel getCaller() {
        return caller;
    }

    public void setCaller(StorageReferenceModel caller) {
    	System.out.println("setting caller to " + caller);
        this.caller = caller;
    }

    protected final String getSignature() {
        return signature;
    }

    public void setSignature(String signature) {
    	System.out.println("setting signature to " + signature);
        this.signature = signature;
    }

    protected final String getChainId() {
        return chainId;
    }

    public void setChainId(String chainId) {
    	System.out.println("setting chain id to " + chainId);
        this.chainId = chainId;
    }

    protected final BigInteger getGasLimit() {
        return gasLimit;
    }

    public void setGasLimit(BigInteger gasLimit) {
        this.gasLimit = gasLimit;
    }

    protected final BigInteger getGasPrice() {
        return gasPrice;
    }

    public void setGasPrice(BigInteger gasPrice) {
        this.gasPrice = gasPrice;
    }
}

File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-js/src/models/requests/NonInitialTransactionRequestModel.ts
import {StorageReferenceModel} from "../values/StorageReferenceModel";
import {TransactionReferenceModel} from "../values/TransactionReferenceModel";
import {TransactionRequestModel} from "./TransactionRequestModel";
import {MarshallingContext} from "../../internal/marshalling/MarshallingContext";
import {HotmokaException} from "../../internal/exception/HotmokaException";

export abstract class NonInitialTransactionRequestModel extends TransactionRequestModel {
    caller: StorageReferenceModel
    nonce: string
    classpath: TransactionReferenceModel
    gasLimit: string
    gasPrice: string

    protected constructor(caller: StorageReferenceModel,
                          nonce: string,
                          classpath: TransactionReferenceModel,
                          gasLimit: string,
                          gasPrice: string) {
        super()

        if (caller === null || caller === undefined) {
            throw new HotmokaException("caller cannot be null or undefined")
        }

        if (gasLimit === null || gasLimit === undefined) {
            throw new HotmokaException("gasLimit cannot be null")
        }

        if (Number(gasLimit) < 0) {
            throw new HotmokaException("gasLimit cannot be negative")
        }

        if (gasPrice === null || gasPrice === undefined) {
            throw new HotmokaException("gasPrice cannot be null")
        }

        if (Number(gasPrice) < 0) {
            throw new HotmokaException("gasPrice cannot be negative")
        }

        if (classpath === null || classpath === undefined) {
            throw new HotmokaException("classpath cannot be null or undefined")
        }

        if (nonce === null || nonce === undefined) {
            throw new HotmokaException("nonce cannot be null")
        }

        if (Number(nonce) < 0) {
            throw new HotmokaException("nonce cannot be negative")
        }

        this.caller = caller
        this.nonce = nonce
        this.classpath = classpath
        this.gasLimit = gasLimit
        this.gasPrice = gasPrice
    }

    protected intoWithoutSignature(context: MarshallingContext): void {
        StorageReferenceModel.intoWithoutSelector(context, this.caller)
        context.writeBigInteger(this.gasLimit)
        context.writeBigInteger(this.gasPrice)
        TransactionReferenceModel.into(context, this.classpath)
        context.writeBigInteger(this.nonce)
    }
}